FreeAPS/Sources/Services/ContactTrick/ContactTrickManager.swift

```diff
@@ -0,0 +1,198 @@
+            return
+        }
+
+        mutableContact.imageData = ContactPicture.getImage(
+            contact: entry,
+            state: state
+        ).pngData()
+
+        let saveRequest = CNSaveRequest()
+        saveRequest.update(mutableContact)
+        do {
+            try contactStore.execute(saveRequest)
+        } catch {
+            print("in updateContact, failed to update the contact")
+        }
+    }
+
+    private func glucoseText(_ glucose: [Readings]) -> (glucose: String, trend: String, delta: String) {
+        let glucoseValue = glucose.first?.glucose ?? 0
+
+        guard !glucose.isEmpty else { return ("--", "--", "--") }
+
+        let delta = glucose.count >= 2 ? glucoseValue - glucose[1].glucose : nil
+
+        let units = settingsManager.settings.units
+        let glucoseText = glucoseFormatter
+            .string(from: Double(
+                units == .mmolL ? Decimal(glucoseValue).asMmolL : Decimal(glucoseValue)
+            ) as NSNumber)!
+
+        let directionText = glucose.first?.direction ?? "↔︎"
+        let deltaText = delta
+            .map {
+                self.deltaFormatter
+                    .string(from: Double(
+                        units == .mmolL ? Decimal($0).asMmolL : Decimal($0)
+                    ) as NSNumber)!
+            } ?? "--"
+
+        return (glucoseText, directionText, deltaText)
+    }
+
+    private func eventualBGString() -> String? {
+        guard let eventualBG = suggestion?.eventualBG else {
+            return nil
+        }
+        let units = settingsManager.settings.units
+        return eventualFormatter.string(
+            from: (units == .mmolL ? eventualBG.asMmolL : Decimal(eventualBG)) as NSNumber
+        )!
+    }
+
+    private var glucoseFormatter: NumberFormatter {
+        let formatter = NumberFormatter()
+        formatter.numberStyle = .decimal
+        formatter.locale = Locale(identifier: "en_US")
+        formatter.maximumFractionDigits = 0
+        if settingsManager.settings.units == .mmolL {
+            formatter.minimumFractionDigits = 1
+            formatter.maximumFractionDigits = 1
+        }
+        formatter.roundingMode = .halfUp
+        return formatter
+    }
+
+    private var eventualFormatter: NumberFormatter {
+        let formatter = NumberFormatter()
+        formatter.numberStyle = .decimal
+        formatter.maximumFractionDigits = 2
+        return formatter
+    }
+
+    private var deltaFormatter: NumberFormatter {
+        let formatter = NumberFormatter()
+        formatter.numberStyle = .decimal
+        formatter.maximumFractionDigits = 1
+        formatter.positivePrefix = "+"
+        return formatter
+    }
+
+    private var suggestion: Suggestion? {
+        storage.retrieve(OpenAPS.Enact.suggested, as: Suggestion.self)
+    }
+}
+
+extension BaseContactTrickManager:
+    GlucoseObserver,
+    SuggestionObserver,
+    SettingsObserver
+{
+    func glucoseDidUpdate(_: [BloodGlucose]) {
```

Comment:
not sure if it's necessary to subscribe to glucose updates, or if subscribing to suggestions is enough

Reply:
I guess BG can also be retrieved from the suggestion, instead of asking `self.coreDataStorage.fetchGlucose(...)`

Reply:
but not `trend` or `delta` 🤔